src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -124,6 +139,61 @@ HNSWIndex<DataType, DistType> *TieredHNSWIndex<DataType, DistType>::getHNSWIndex
     return reinterpret_cast<HNSWIndex<DataType, DistType> *>(this->index);
 }
 
+template <typename DataType, typename DistType>
+int TieredHNSWIndex<DataType, DistType>::deleteLabelFromHNSW(labelType label) {
+    auto *hnsw_index = getHNSWIndex();
+
+    // Get the required data about the relevant ids to delete.
+    // Internally, this will hold the index data lock.
+    auto internal_ids = hnsw_index->markDelete(label);
+
+    if (internal_ids.empty()) {
+        // Label doesn't exist, or it has already marked as deleted - we can finish now.
+        return 0;
+    }
+    vecsim_stl::vector<size_t> ids_top_level(this->allocator);
+    for (idType id : internal_ids) {
+        ids_top_level.push_back(hnsw_index->getElementTopLevel(id));
+    }
+
+    for (size_t i = 0; i < internal_ids.size(); i++) {
+        idType id = internal_ids[i];
+        vecsim_stl::vector<HNSWRepairJob *> repair_jobs(this->allocator);
+        auto *swap_job = new (this->allocator) HNSWSwapJob(this->allocator, id);
+
+        // Go over all the deleted element links in every level and create repair jobs.
+        auto incoming_edges = hnsw_index->safeCollectAllNodeIncomingNeighbors(id, ids_top_level[i]);
+        {
+            // Protect the id->repair_jobs lookup while we update it with the new jobs.
+            std::unique_lock<std::mutex> lock(this->idToRepairJobsGuard);
+            for (pair<idType, ushort> &node : incoming_edges) {
+                auto *repair_job =
+                    new (this->allocator) HNSWRepairJob(this->allocator, node.first, node.second,
+                                                        executeRepairJobWrapper, this, swap_job);
+                repair_jobs.emplace_back(repair_job);
+                // Insert the repair job into the repair jobs lookup (for fast update in case that
+                // the node id is changed due to swap job).
+                if (idToRepairJobs.find(node.first) == idToRepairJobs.end()) {
+                    idToRepairJobs.insert(
+                        {node.first, vecsim_stl::vector<HNSWRepairJob *>(this->allocator)});
+                }
+                idToRepairJobs.at(node.first).push_back(repair_job);
```

Comment:
can't we use this map to decide if we need a new job? an id doesn't really need more than one repair at a time

Reply:
We can, but there is another reason why it's better (IMO) to have another repair job anyway - and that is for tracking the repair jobs that correspond a swap job (let's discuss it)